js/passwordManager/passwordManager.js

```diff
@@ -0,0 +1,249 @@
+// Bitwarden password manager. Requires session key to unlock the vault.
+class Bitwarden {
+  constructor() {
+    this.sessionKey = null
+    this.lastCall = null
+    this.name = 'Bitwarden'
+  }
+
+  // Checks if given command runs and produces output.
+  async _checkCommand(command, args) {
+    try {
+      let process = new ProcessSpawner(command, args)
+      let data = await process.execute()
+      if (data.length > 0) {
+        return true
+      }
+    } catch (ex) {
+      return false
+    }
+  }
+
+  // Returns a Bitwarden-CLI tool path by checking possible locations.
+  // First it checks if the tool was installed for Min specifically by
+  // by checking the settings value. If that is not set or doesn't point
+  // to a valid executable, it check the if 'bw' is available globally.
+  async _getToolPath() {
+    let localPath = settings.get('bitwardenPath')
+    if (localPath) {
+      let local = false;
+      try {
+        await fs.promises.access(localPath, fs.constants.X_OK)
+        local = true;
+      } catch (e) {}
+      if (local) {
+        return localPath
+      }
+    }
+
+    let global;
+    if (platformType === "windows") {
+      global = await this._checkCommand('where', ['bw'])
+    } else {
+      global = await this._checkCommand('which', ['bw'])
+    }
+  
+    if (global) {
+      return 'bw'
+    }
+
+    return null
+  }
+
+  // Checks if Bitwarden integration is configured properly by trying to
+  // obtain a valid Bitwarden-CLI tool path.
+  async checkIfConfigured() {
+    this.path = await this._getToolPath()
+    return this.path != null
+  }
+
+  // Returns current Bitwarden-CLI status. If we have a session key, then
+  // password store is considered unlocked.
+  isUnlocked() {
+    return this.sessionKey != null
+  }
+
+  // Tries to get a list of credential suggestions for a given domain name.
+  // If password store is locked, the method will try to unlock it by
+  async getSuggestions(domain) {
+    if (this.lastCall != null) {
+      return this.lastCall
+    }
```

Comment:
What's the reason for this?

If I'm understanding correctly, if two websites request passwords at almost the same time, the second website could end up with the passwords for the first website, which would be fairly bad.

Reply:
The idea is to guard from stacking up multiple unlock requests/dialogs in case when user activates the autofill multiple times. By accidently double-clicking on the "fill" icon, for example.

Reply:
OK, that makes sense, I think this can be fixed by keeping track of lastCalll separately per-domain then.